api/task_manager.cc

```diff
@@ -0,0 +1,167 @@
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        sharded<task_manager>& tm = ctx.tm;
+        try {
+            task_manager::foreign_task_ptr task = co_await task_manager::lookup_task_on_all_shards(tm, id);
+            task = co_await tm.invoke_on(task.get_owner_shard(), [task = std::move(task)] (auto& tm) mutable -> future<task_manager::foreign_task_ptr> {
+                auto state = task->get_status().state;
+                if (state == task_manager::task_state::done || state == task_manager::task_state::failed) {
+                    task->unregister_task();
+                }
+                co_return std::move(task);
+            });
+            co_return co_await retrieve_status(std::move(task));
+        } catch (...) {
+            co_return coroutine::exception(std::current_exception());
+        }
+
+    });
+
+    tm::abort_task.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        sharded<task_manager>& tm = ctx.tm;
+        try {
+            task_manager::foreign_task_ptr task = co_await task_manager::lookup_task_on_all_shards(tm, id);
+            co_await tm.invoke_on(task.get_owner_shard(), [t = std::move(task)] (auto& tm) mutable -> future<> {
+                task_manager::task_ptr task = t.release();
+                if (!task->is_abortable()) {
+                    co_await coroutine::return_exception(std::runtime_error("Requested task cannot be aborted"));
+                }
+                co_await task->abort();
+            });
+            co_return json_void();
+        } catch (...) {
+            co_return coroutine::exception(std::current_exception());
+        }
+    });
+
+    tm::wait_task.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        sharded<task_manager>& tm = ctx.tm;
+        try {
+            task_manager::foreign_task_ptr task = co_await task_manager::lookup_task_on_all_shards(tm, id);
+            task = co_await tm.invoke_on(task.get_owner_shard(), [t = std::move(task)] (auto& tm) mutable -> future<task_manager::foreign_task_ptr> {
+                task_manager::task_ptr task = t.release();
+                return task->done().then_wrapped([task] (auto f) {
+                    task->unregister_task();
+                    if (f.failed()) {
+                        auto ex = f.get_exception();
+                        return make_exception_future<task_manager::foreign_task_ptr>(std::move(ex));
+                    }
+                    return make_ready_future<task_manager::foreign_task_ptr>(make_foreign<task_manager::task_ptr>(task));
+                });
+            });
+            co_return co_await retrieve_status(std::move(task));
+        } catch (...) {
+            co_return coroutine::exception(std::current_exception());
```

Comment:
Isn't that redundant?
Wouldn't that happen automatically without try/catch?

Reply:
I see it's a repeated pattern, and I'm not sure why it's needed.
The coroutine wrapper that the compiler generates catches any exceptions and turns them into an exceptional future returned by the coroutine.